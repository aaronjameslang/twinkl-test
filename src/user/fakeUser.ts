// eslint-disable-next-line import/no-extraneous-dependencies
import { faker } from '@faker-js/faker';
import { randomUUID } from 'crypto';
import { UserInput, UserOutput, UserRecord } from './User';
import { hashPassword } from './hashPassword';

interface UserFake {
  input: UserInput
  output: UserOutput
  getRecord: () => Promise<UserRecord>
}

export function fakeUser(): UserFake {
  const id = randomUUID();
  const input: UserInput = {
    created: faker.date.recent().toISOString(),
    email: faker.internet.email(),
    name: faker.person.fullName(),
    password: faker.internet.password() + 1,
    type: faker.helpers.arrayElement([
      'parent',
      'student',
      'teacher',
      'tutor',
    ]),
  };
  const output = { ...input, id, password: undefined };
  async function getRecord() {
    const { digest, salt } = await hashPassword(input.password);
    return { ...output, digest, salt };
  }
  return {
    input,
    output,
    getRecord,
  };
}
