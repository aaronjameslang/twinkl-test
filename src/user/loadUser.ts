import { existsSync, readFileSync } from 'fs';
import { UserRecord } from './User';

export function loadUser(id: string): UserRecord | null {
  const filename = `./data/users/${id}.json`;
  if (!existsSync(filename)) return null;
  const json = readFileSync(filename, 'utf-8');
  return JSON.parse(json);
}
