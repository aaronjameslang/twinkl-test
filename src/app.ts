import bodyParser from 'body-parser';
import express, { Express, Request, Response } from 'express';
import { handleZodError } from './handleZodError';
import { getUser } from './user/getUser';
import { postUser } from './user/postUser';

export const app: Express = express();

app.use(bodyParser.json());

app.get('/', (req: Request, res: Response) => {
  res.send('Hello World!');
});

app.get('/user/:id', getUser);
app.post('/user', postUser);

app.use(handleZodError);
