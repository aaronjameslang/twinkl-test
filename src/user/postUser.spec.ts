import request from 'supertest';
import { app } from '../app';
import { fakeUser } from './fakeUser';
import { loadUser } from './loadUser';

describe('POST /user', () => {
  it('should return 201', async () => {
    await request(app)
      .post('/user')
      .send(fakeUser().input)
      .expect(201);
  });
  it('should return user ID', async () => {
    const response = await request(app)
      .post('/user')
      .send(fakeUser().input);
    const { id } = response.body;
    expect(typeof id).toBe('string');
    expect(id).toHaveLength(36);
  });
  it('should save user', async () => {
    const { input } = fakeUser();
    const response = await request(app)
      .post('/user')
      .send(input);
    const { id } = response.body;
    const actualRecord = loadUser(id);
    const actualOutput = { ...actualRecord, digest: undefined, salt: undefined };
    const expectedOutput = { ...input, id, password: undefined };
    expect(actualOutput).toEqual(expectedOutput);
  });

  // --- Unhappy Paths ---
  Object.keys(fakeUser().input).forEach((key) => {
    ['', null, undefined].forEach((value) => {
      const user = fakeUser().input;
      // @ts-expect-error
      user[key] = value;
      const description = `user.${key} = ${JSON.stringify(value)}`;
      it(`should return 422 for ${description}`, async () => {
        await request(app)
          .post('/user')
          .send(user)
          .expect(422);
      });
      it(`should return error details for ${description}`, async () => {
        const response = await request(app)
          .post('/user')
          .send(user);
        expect(response.body).toMatchSnapshot();
      });
    });
  });
});
