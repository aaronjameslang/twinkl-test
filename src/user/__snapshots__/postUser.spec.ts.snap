// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`POST /user should return error details for user.created = "" 1`] = `
{
  "zod": [
    {
      "code": "too_small",
      "exact": false,
      "inclusive": true,
      "message": "String must contain at least 1 character(s)",
      "minimum": 1,
      "path": [
        "created",
      ],
      "type": "string",
    },
  ],
}
`;

exports[`POST /user should return error details for user.created = null 1`] = `
{
  "zod": [
    {
      "code": "invalid_type",
      "expected": "string",
      "message": "Expected string, received null",
      "path": [
        "created",
      ],
      "received": "null",
    },
  ],
}
`;

exports[`POST /user should return error details for user.created = undefined 1`] = `
{
  "zod": [
    {
      "code": "invalid_type",
      "expected": "string",
      "message": "Required",
      "path": [
        "created",
      ],
      "received": "undefined",
    },
  ],
}
`;

exports[`POST /user should return error details for user.email = "" 1`] = `
{
  "zod": [
    {
      "code": "invalid_string",
      "message": "Invalid email",
      "path": [
        "email",
      ],
      "validation": "email",
    },
  ],
}
`;

exports[`POST /user should return error details for user.email = null 1`] = `
{
  "zod": [
    {
      "code": "invalid_type",
      "expected": "string",
      "message": "Expected string, received null",
      "path": [
        "email",
      ],
      "received": "null",
    },
  ],
}
`;

exports[`POST /user should return error details for user.email = undefined 1`] = `
{
  "zod": [
    {
      "code": "invalid_type",
      "expected": "string",
      "message": "Required",
      "path": [
        "email",
      ],
      "received": "undefined",
    },
  ],
}
`;

exports[`POST /user should return error details for user.name = "" 1`] = `
{
  "zod": [
    {
      "code": "too_small",
      "exact": false,
      "inclusive": true,
      "message": "String must contain at least 1 character(s)",
      "minimum": 1,
      "path": [
        "name",
      ],
      "type": "string",
    },
  ],
}
`;

exports[`POST /user should return error details for user.name = null 1`] = `
{
  "zod": [
    {
      "code": "invalid_type",
      "expected": "string",
      "message": "Expected string, received null",
      "path": [
        "name",
      ],
      "received": "null",
    },
  ],
}
`;

exports[`POST /user should return error details for user.name = undefined 1`] = `
{
  "zod": [
    {
      "code": "invalid_type",
      "expected": "string",
      "message": "Required",
      "path": [
        "name",
      ],
      "received": "undefined",
    },
  ],
}
`;

exports[`POST /user should return error details for user.password = "" 1`] = `
{
  "zod": [
    {
      "code": "too_small",
      "exact": false,
      "inclusive": true,
      "message": "String must contain at least 8 character(s)",
      "minimum": 8,
      "path": [
        "password",
      ],
      "type": "string",
    },
    {
      "code": "invalid_string",
      "message": "Password must contain a number",
      "path": [
        "password",
      ],
      "validation": "regex",
    },
    {
      "code": "invalid_string",
      "message": "Password must contain an uppercase letter",
      "path": [
        "password",
      ],
      "validation": "regex",
    },
    {
      "code": "invalid_string",
      "message": "Password must contain a lowercase letter",
      "path": [
        "password",
      ],
      "validation": "regex",
    },
  ],
}
`;

exports[`POST /user should return error details for user.password = null 1`] = `
{
  "zod": [
    {
      "code": "invalid_type",
      "expected": "string",
      "message": "Expected string, received null",
      "path": [
        "password",
      ],
      "received": "null",
    },
  ],
}
`;

exports[`POST /user should return error details for user.password = undefined 1`] = `
{
  "zod": [
    {
      "code": "invalid_type",
      "expected": "string",
      "message": "Required",
      "path": [
        "password",
      ],
      "received": "undefined",
    },
  ],
}
`;

exports[`POST /user should return error details for user.type = "" 1`] = `
{
  "zod": [
    {
      "code": "invalid_enum_value",
      "message": "Invalid enum value. Expected 'parent' | 'student' | 'teacher' | 'tutor', received ''",
      "options": [
        "parent",
        "student",
        "teacher",
        "tutor",
      ],
      "path": [
        "type",
      ],
      "received": "",
    },
  ],
}
`;

exports[`POST /user should return error details for user.type = null 1`] = `
{
  "zod": [
    {
      "code": "invalid_type",
      "expected": "'parent' | 'student' | 'teacher' | 'tutor'",
      "message": "Expected 'parent' | 'student' | 'teacher' | 'tutor', received null",
      "path": [
        "type",
      ],
      "received": "null",
    },
  ],
}
`;

exports[`POST /user should return error details for user.type = undefined 1`] = `
{
  "zod": [
    {
      "code": "invalid_type",
      "expected": "'parent' | 'student' | 'teacher' | 'tutor'",
      "message": "Required",
      "path": [
        "type",
      ],
      "received": "undefined",
    },
  ],
}
`;
